gossip/client.go

```diff
@@ -147,28 +156,59 @@ func (c *client) gossip(g *Gossip, stopper *stop.Stopper) error {
 				log.Infof("gossip: received %d info(s) from %s", infoCount, c.addr)
 			}
 		}
-		g.mu.Lock()
-		c.peerID = reply.NodeID
-		g.outgoing.addNode(c.peerID)
 		freshCount := g.is.combine(reply.Delta, reply.NodeID)
 		if freshCount > 0 {
-			c.lastFresh = now
+			log.Infof("gossip: %d info(s) were fresh from %s", freshCount, c.addr)
 		}
-		remoteMaxSeq = reply.MaxSeq
+	}
+	c.peerID = reply.NodeID
+	g.outgoing.addNode(c.peerID)
+	c.remoteHighWaterStamps = reply.HighWaterStamps
 
-		// If we have the sentinel gossip, we're considered connected.
-		g.checkHasConnected()
-		g.mu.Unlock()
+	// If we have the sentinel gossip, we're considered connected.
+	g.checkHasConnected()
+	return nil
+}
 
-		// Check whether this outgoing client is duplicating work already
-		// being done by an incoming client. To avoid mutual shutdown, we
-		// only shutdown our client if our node ID is less than the peer's.
-		if g.hasIncoming(c.peerID) && nodeID < c.peerID {
-			return util.Errorf("stopping outgoing client %d @ %s; already have incoming", c.peerID, c.addr)
-		}
-		// Check whether peer node is too boring--disconnect if yes.
-		if nodeID != c.peerID && (now-c.lastFresh) > int64(maxWaitForNewGossip) {
-			return util.Errorf("peer is too boring")
+// gossip loops, sending deltas of the infostore and receiving deltas
+// in turn. If an alternate is proposed on response, the client addr
+// is modified and method returns for forwarding by caller.
+func (c *client) gossip(g *Gossip, stopper *stop.Stopper) error {
+	g.mu.Lock()
+	nodeID := g.is.NodeID
+	addr := util.MakeUnresolvedAddr(g.is.NodeAddr.Network(), g.is.NodeAddr.String())
+	g.mu.Unlock()
+
+	lAddr := util.MakeUnresolvedAddr(c.rpcClient.LocalAddr().Network(), c.rpcClient.LocalAddr().String())
+	getGossipCall := c.getGossip(g, nodeID, addr, lAddr)
+	sendGossipCall := &netrpc.Call{}
+
+	// Loop until stopper is signalled, or until either the gossip or
+	// RPC clients are closed. The getGossipCall is a hanging get,
+	// returning results only when the remote server has new gossip
+	// information to share. The sendGossipCall is sent to the remote
+	// server when this node has new gossip information to share with
+	// the server.
+	for {
+		select {
```

Comment:
You mean where did it go? I'm figuring we just skip those here. Other parts of the system are already monitoring our links for failures and timeouts and will kill them. If the rpc is closed out from under the gossip client, it does the right thing. The old timeout code was redundant.